### PR TITLE
Transitioning from Python 2.7 to Python 3

### PR DESCRIPTION
**Pull request by jcfr**
See https://www.slicer.org/wiki/Documentation/Labs/Slicer5-roadmap#Python3

### CODE REVIEW

**Comment by jcfr** [comment]
Shorty, this topic will be updated. In the mean time, Slicer can be started with Python 3.

image

See https://discourse.slicer.org/t/updating-slicer-to-work-with-python-3/4662/4?u=jcfr

File: Modules/Loadable/Segmentations/EditorEffects/Python/SegmentEditorHollowEffect.py [hunk 1]
```diff
@@ -154,7 +154,7 @@ def onApply(self):
     kernelSizePixel = self.getKernelSizePixel()
     if shellMode == MEDIAL_SURFACE:
       # both erosion and dilation will be applied, so kernel size must be half on each side
-      kernelSizePixel = [kernelSizePixel[0]/2, kernelSizePixel[1]/2, kernelSizePixel[2]/2]
+      kernelSizePixel = [int(kernelSizePixel[0],2), int(kernelSizePixel[1],2), int(kernelSizePixel[2],2)]
```
Comment (jcfr) [inline Modules/Loadable/Segmentations/EditorEffects/Python/SegmentEditorHollowEffect.py]: @lassoan  Would be great if you could review the changes associated with the Segment editor effects.

Comment (jcfr) [reply Modules/Loadable/Segmentations/EditorEffects/Python/SegmentEditorHollowEffect.py]: The gist is that in `1/3` returns `0` in python 2 and returns `0.3333333` in python 3

**Comment by jcfr** [comment]
Error on macOS:

```
[100%] Relocate _sysconfigdata_m_darwin_.py and update pybuilddir.txt
dyld: lazy symbol binding failed: Symbol not found: _clock_gettime
  Referenced from: /Volumes/Dashboards/Experimental/jc/Slicer-Release-UnixMakefiles/python-build/lib/libpython3.6m.dylib
  Expected in: /usr/lib/libSystem.B.dylib

dyld: Symbol not found: _clock_gettime
  Referenced from: /Volumes/Dashboards/Experimental/jc/Slicer-Release-UnixMakefiles/python-build/lib/libpython3.6m.dylib
  Expected in: /usr/lib/libSystem.B.dylib

```

**Comment by jcfr** [comment]
Error on macOS has been fixed in https://github.com/python-cmake-buildsystem/python-cmake-buildsystem/pull/246

**Comment by jcfr** [comment]
Windows issue has been fixed in . Module `` introduced in Python 3.6 is now build.

macOS package was successfully created ! Slicer starts but there are few minor issues reported:

```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/kitware/Desktop/Slicer.app/Contents/lib/Python/lib/python3.6/imp.py", line 170, in load_source
    module = _exec(spec, sys.modules[name])
  File "<frozen importlib._bootstrap>", line 618, in _exec
  File "<frozen importlib._bootstrap_external>", line 674, in exec_module
  File "<frozen importlib._bootstrap_external>", line 781, in get_code
  File "<frozen importlib._bootstrap_external>", line 741, in source_to_code
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/Users/kitware/Desktop/Slicer.app/Contents/lib/Slicer-4.11/qt-scripted-modules/SimpleFilters.py", line 228
    print "\n".join(printStr)
             ^
SyntaxError: invalid syntax
loadSourceAsModule - Failed to load file "/Users/kitware/Desktop/Slicer.app/Contents/lib/Slicer-4.11/qt-scripted-modules/SimpleFilters.py"  as module "SimpleFilters" !
Fail to instantiate module  "SimpleFilters"
Switch to module:  "Welcome"
Traceback (most recent call last):
  File "/Users/kitware/Desktop/Slicer.app/Contents/lib/Slicer-4.11/qt-scripted-modules/DICOM.py", line 76, in performPostModuleDiscoveryTasks
    DICOMLib.setDatabasePrecacheTags()
  File "/Users/kitware/Desktop/Slicer.app/Contents/lib/Slicer-4.11/qt-scripted-modules/DICOMLib/DICOMWidgets.py", line 41, in setDatabasePrecacheTags
    plugin = slicer.modules.dicomPlugins[pluginClass]()
  File "/Users/kitware/Desktop/Slicer.app/Contents/lib/Slicer-4.11/qt-scripted-modules/MultiVolumeImporterPlugin.py", line 57, in __init__
    for tagName,tagVal in self.multiVolumeTags.iteritems():
AttributeError: 'dict' object has no attribute 'iteritems'
```

**Comment by jcfr** [comment]
Closing. Corresponding changes have been integrated in [r28100](http://viewvc.slicer.org/viewvc.cgi/Slicer4?view=revision&revision=28100) to [r28129](http://viewvc.slicer.org/viewvc.cgi/Slicer4?view=revision&revision=28129)

**Comment by cpinter** [comment]
I tried building the latest revision using VS2019 (2015 toolset), and I got this when the superbuild was configuring CTK:

> LINK : fatal error LNK1104: cannot open file 'C:\d\S4D\python-install\libs\python36.lib' [C:\d\S4D\CTK-build\PythonQt-build\PythonQt.vcxproj] [C:\d\S4D\CTK-build\PythonQt.vcxproj]
    Performing update step for 'QtTesting'

This was my first build with 2019, but the error is about Python3, and I heard that Slicer builds well using 2019, so thought this could be related to these latest changes.

**Comment by lassoan** [comment]
A clean build with VS2017 (VS2015 toolset) failed for me on Windows, too:

```
18>ClCompile:
18>  C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\bin\x86_amd64\CL.exe /c /I"C:\D\S4R\Python-3.6.7\Include" /I"C:\D\S4R\Python-3.6.7\PC" /I"C:\D\S4R\Python-3.6.7\Python" /nologo /W2 /WX- /O2 /Ob1 /D WIN32 /D _WINDOWS /D NDEBUG /D Py_BUILD_CORE /D "CMAKE_INTDIR=\"Release\"" /D _MBCS /Gm- /MD /GS /fp:precise /Zc:wchar_t /Zc:forScope /Zc:inline /Fo"_freeze_importlib.dir\Release\\" /Fd"_freeze_importlib.dir\Release\vc140.pdb" /Gd /TC /errorReport:queue  /Zm200 "C:\D\S4R\Python-3.6.7\Modules\_io\_iomodule.c"
18>  _iomodule.c
18>C:\Program Files (x86)\Windows Kits\10\Include\10.0.17134.0\um\winnt.h(154): fatal error C1189: #error:  "No Target Architecture" [C:\D\S4R\python-build\CMakeBuild\libpython\_freeze_importlib.vcxproj]
18>Done Building Project "C:\D\S4R\python-build\CMakeBuild\libpython\_freeze_importlib.vcxproj" (default targets) -- FAILED.

```

See full log here: https://1drv.ms/t/s!Arm_AFxB9yqHt54qOQ3phL4bGFYloQ

@jcfr Do you have any advice?

**Comment by jcfr** [comment]
I tested with VS2015. I will look into vs2017, I think there is already a
PR on python-cmake-buildsystem addressing the _iomodule failure.

**Comment by lassoan** [comment]
Thanks for the information. VS2017 is just the IDE, the compiler is VS2015. I built Slicer master branch yesterday on the same computer with the same build options without any problems.

**Comment by lassoan** [comment]
Replacing `#include <consoleapi.h>` by `#include <windows.h>` in C:\D\S4R\Python-3.6.7\Modules\_io\_iomodule.c seems to fix the problem. It makes sense, as in general it is not allowed to include internal windows headers in applications but only the top-level windows.h header.

**Comment by jcfr** [comment]
> A clean build with VS2017 (VS2015 toolset) failed for me on Windows

@lassoan  This is fixed in [r28138](http://viewvc.slicer.org/viewvc.cgi/Slicer4?view=revision&revision=28138)

**Comment by jcfr** [comment]
> tried building the latest revision using VS2019 (2015 toolset) [... ] There is no python36.lib, but there is a python27.lib in that directory.

Could you confirm this was a clean build ?

**Comment by lassoan** [comment]
There are many small issues related to python 2 to 3 switch. Where we should track them? Adding a mantis issue for each would be too much. Should we use a wiki page, online spreadsheet, discourse topic,...?

**Comment by jcfr** [comment]
> Should we use a wiki page

so far, I have been updating this wiki page: https://www.slicer.org/wiki/Documentation/Labs/Slicer5-roadmap#Python3

**Comment by cpinter** [comment]
> Could you confirm this was a clean build ?

Yes it was.

I'll try again from scratch with your new fixes.

**Comment by jcfr** [comment]
> > Could you confirm this was a clean build ?

> Yes it was.
> I'll try again from scratch with your new fixes.

The fix I integrated only fixes a compilation issue associated with the `iomodule`, it is strange that a library named `python27.lib` was created.

Could you check that you are not explicitly setting variable in the Slicer cache. For example, from with a `.bat` script.

**Comment by cpinter** [comment]
It is indeed strange that a 2.7 lib was created in a clean build!

This is my configure command:

> cmake -G "Visual Studio 16 2019" -T "v140" -DQt5_DIR=C:/Qt/Qt5.10.0/5.10.0/msvc2015_64/lib/cmake/Qt5 -DSlicer_USE_SimpleITK:BOOL=ON -DSlicer_USE_PYTHONQT_WITH_OPENSSL:BOOL=ON c:/d/Slicer

**Comment by jcfr** [comment]
> cmake -G "Visual Studio 16 2019" -T "v140"

worth noting that you are not selecting the Win64 generator

Waiting we update the build system to pass [CMAKE_GENERATOR_PLATFORM](https://cmake.org/cmake/help/latest/variable/CMAKE_GENERATOR_PLATFORM.html) around (itself initialized by invoking cmake with `-A x64`), you should specify `Visual Studio 16 2019 Win64` instead.



May be worth trying to configure and build only python-cmake-buildsystem:

```
git clone git://github.com/python-cmake-buildsystem/python-cmake-buildsystem
mkdir python-cmake-buildsystem-build
cd python-cmake-buildsystem-build
cmake -G "Visual Studio 16 2019" -T "v140" -A x64 -DBUILD_LIBPYTHON_SHARED:BOOL=1 ../python-cmake-buildsystem
cmake --build . --config Release -- /m
```

**Comment by cpinter** [comment]
In the new CMake versions it seems that if you omit the configuration the default will be x64.

This is why I think it will be:
image

**Comment by cpinter** [comment]
Also I can confirm that Visual Studio is actually building Slicer in 64 bits.

**Comment by jcfr** [comment]
Since i don't know if the initialization to default value happen in the gui dialog or in the regular CMake code, I don't know if the platform is correctly initialized by default in sub project. Could you check that  `CMAKE_GENERATOR_PLATFORM ` is correctly set in a sub project ?

**Comment by jcfr** [comment]
> re: restart() issue

@lassoan  This is related to the auto-completion and is independent of the function being executed. If the function is auto-completed, it will not be evaluated.

**Comment by cpinter** [comment]
CMAKE_GENERATOR_PLATFORM is empty.

But as I said Visual Studio is building in 64 bits.
Also
CMAKE_CXX_COMPILER:FILEPATH=C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/amd64/cl.exe
CMAKE_CXX_COMPILER:FILEPATH=C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/amd64/cl.exe
CMAKE_C_COMPILER:FILEPATH=C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/amd64/cl.exe

**Comment by cpinter** [comment]
@jcfr This time the build succeeded.

I am 100% sure that it was a clean build for the first time, and that python27 thing puzzles me too. If it happens again I'll let you know. Thanks!

**Comment by jcfr** [comment]
Excellent news :)

**Comment by jcfr** [comment]
> > re: restart() issue

> @lassoan  This is related to the auto-completion and is independent of the function being executed. If 
> the function is auto-completed, it will not be evaluated. 

This will be fixed by https://github.com/commontk/CTK/pull/859

**Pull request closed without merging**